broadinstitute/single_cell_portal_core
# Client-side file validation for remote files in sync UI (SCP-4017)

This adds in-app file validation to the sync UI.  It enables users to fix problems in their files drastically faster.

Previously, when importing files from Terra workspaces into SCP, users had to wait at least 2-3 minutes to learn of any validation errors.  And those errors were emailed to the user.  Now, users learn of such sync file validation problems in < 10 seconds and right in the web app.

Sync client-side file validation (CSFV) extends similar functionality recently introduced to the upload UI.  The validations are the same.  Other aspects of the sync UI still use the legacy design, as a larger sync UI revamp is being contemplated.  The main technical difference between upload and sync CSFV is that the latter uses [HTTP range requests](https://developer.mozilla.org/en-US/docs/Web/HTTP/Range_requests) to read the first part of the file.  That's because sync files are remote, existing in a Google Cloud Storage bucket, unlike upload files which exist on a user's local machine.  Relevant new Mixpanel analytics events are also recorded, e.g. whether the full file was fetched and how much time that took.

Here's how it looks:

https://user-images.githubusercontent.com/1334561/153594323-bb45dcf8-cfd4-4e89-94b2-47f9998cc766.mov

To test:
* Add files to a study's Google bucket, including valid and invalid files, both larger and smaller than 50 MiB
* Go to Sync page for study
* Select file type from each file's dropdown menu
* Confirm no issues are reported for small valid files
* Confirm errors are reported for small invalid files
* Confirm only "incomplete" warning is reported for large valid files, e.g. [Dense_matrix_204_MB_good.txt](https://console.cloud.google.com/storage/browser/_details/broad-singlecellportal-public/test/Dense_matrix_204_MB_good.txt)
* Confirm errors and "incomplete" warning are reported for large invalid files, e.g. [Dense_matrix_204_MB_bad_GENE_header.txt](https://console.cloud.google.com/storage/browser/_details/broad-singlecellportal-public/test/Dense_matrix_204_MB_bad_GENE_header.txt)
* Confirm click "Sync" button still works

This satisfies SCP-4017.
